crates/client/eth/src/client.rs

```diff
@@ -0,0 +1,164 @@
+    pub l1_core_contract: StarknetCoreContractInstance<Http<Client>, RootProvider<Http<Client>>>,
+}
+
+impl EthereumClient {
+    /// Create a new EthereumClient instance with the given RPC URL
+    pub async fn new(url: Url, l1_core_address: Address) -> anyhow::Result<Self> {
+        let provider = ProviderBuilder::new().on_http(url);
+        let core_contract = StarknetCoreContract::new(l1_core_address, provider.clone());
+
+        Ok(Self { provider: Arc::new(provider), l1_core_contract: core_contract })
+    }
+
+    /// Retrieves the latest Ethereum block number
+    pub async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        let block_number = self.provider.get_block_number().await?.as_u64();
+        Ok(block_number)
+    }
+
+    /// Get the block number of the last occurrence of a given event.
+    pub async fn get_last_event_block_number<T: SolEvent>(&self) -> anyhow::Result<u64> {
+        let latest_block: u64 = self.get_latest_block_number().await?;
+
+        // Assuming an avg Block time of 15sec we check for a LogStateUpdate occurence in the last ~24h
+        let filter = Filter::new()
+            .from_block(latest_block - 6000)
+            .to_block(latest_block)
+            .address(*self.l1_core_contract.address());
+
+        let logs = self.provider.get_logs(&filter).await?;
+
+        let filtered_logs = logs.into_iter().filter_map(|log| log.log_decode::<T>().ok()).collect::<Vec<_>>();
+
+        if let Some(last_log) = filtered_logs.last() {
+            let last_block: u64 = last_log.block_number.context("no block number in log")?;
+            Ok(last_block)
+        } else {
+            bail!("no event found")
+        }
+    }
+
+    /// Get the last Starknet block number verified on L1
+    pub async fn get_last_verified_block_number(&self) -> anyhow::Result<u64> {
+        let block_number = self.l1_core_contract.stateBlockNumber().call().await?;
+        let last_block_number: u64 = (block_number._0).as_u64();
+        Ok(last_block_number)
+    }
+
+    /// Get the last Starknet state root verified on L1
+    pub async fn get_last_state_root(&self) -> anyhow::Result<StarkFelt> {
+        let state_root = self.l1_core_contract.stateRoot().call().await?;
+        u256_to_starkfelt(state_root._0)
+    }
+
+    /// Get the last Starknet block hash verified on L1
+    pub async fn get_last_verified_block_hash(&self) -> anyhow::Result<StarkFelt> {
+        let block_hash = self.l1_core_contract.stateBlockHash().call().await?;
+        u256_to_starkfelt(block_hash._0)
+    }
+}
+
+#[cfg(test)]
+pub mod eth_client_getter_test {
+    use super::*;
+    use alloy::node_bindings::{Anvil, AnvilInstance};
+    use alloy::primitives::U256;
+    use rstest::*;
+    use tokio;
+
+    // https://etherscan.io/tx/0xcadb202495cd8adba0d9b382caff907abf755cd42633d23c4988f875f2995d81#eventlog
+    // The txn we are referring to it is here ^
+    const ETH_URL: &str = "https://eth.merkle.io";
```

Comment:
I don't know what to feel about putting an rpc provider in our tests
An env var might be better what do you think?

Reply:
yeah sounds good, one more thing that needs to added is secret in the github for the workflow. Because tests depends on it.

Reply:
added `.env.test` for this